fix(iam): validate PolicyStatement SID is alphanumeric for identity policies

camerondurham opened the pull request:
Resolves #34819 by throwing error when PolicyStatement Statement id contains characters not allowed by the API.

Note this _does_ duplicate #34828. I realized that PR makes too many changes to existing integ tests. This change is intended to be less intrusive.

Reference: https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_sid.html
>The Sid element supports ASCII uppercase letters (A-Z), lowercase letters (a-z), and numbers (0-9).


### Issue # (if applicable)

Closes #34819

### Reason for this change

CDK build allows users to write invalid statement ids containing dashes when building policy statements, then failed on deployment. To reduce time to a successful deployment, I believe the CDK should throw an exception at build time to help user make a fix locally before a failed deployment.


### Description of changes
Validate that statement id matches requirements specified in https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_sid.html in the constructor.

Add tests that verify only valid statement ids are set in constructor.

### Describe any new or updated permissions being added
n/a


### Description of how you validated changes

yes, added unit tests

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

camerondurham commented:
Will fix later today after work.

vishaalmehrishi commented on packages/aws-cdk-lib/cx-api/lib/features.ts:
```diff
@@ -1433,6 +1434,20 @@ export const FLAGS: Record<string, FlagInfo> = {
+      that may use invalid SID characters.`,
+    introducedIn: { v2: 'V2NEXT' },
+    recommendedValue: true,
+  },
```
This feature flag definition is missing the `compatibilityWithOldBehaviorMd` field. All other `FlagType.BugFix` flags include this field - for example, the adjacent `IAM_OIDC_REJECT_UNAUTHORIZED_CONNECTIONS` flag at line 1434 has `compatibilityWithOldBehaviorMd: 'Disable the feature flag to allow unsecure OIDC connection.'`.

This field is used to generate the "Compatibility with old behavior" section in `FEATURE_FLAGS.md` and tells users how to revert if needed.

```typescript
compatibilityWithOldBehaviorMd: 'Set this flag to `false` to disable SID validation and allow non-alphanumeric characters in PolicyStatement SIDs.',
```

vishaalmehrishi commented on packages/@aws-cdk-testing/framework-integ/test/aws-iam/test/integ.policy-statement-sid.ts:
```diff
@@ -0,0 +1,33 @@
+}
+
+const app = new App();
+new IntegTest(app, 'iam-policy-statement-sid', {
```
This integration test doesn't enable the `@aws-cdk/aws-iam:policyStatementValidateSid` feature flag. Without it, the validation code path is never exercised during the integ test - it only proves that valid SIDs deploy (which was never in question).

Let's enable the flag so the test verifies that validation doesn't interfere with valid deployments:

```typescript
const app = new App({
  context: {
    '@aws-cdk/aws-iam:policyStatementValidateSid': true,
  },
});
```

camerondurham replied:
My bad, good point yes I totally missed this. Updating in latest commit.

vishaalmehrishi commented on CHANGELOG.md:
```diff
@@ -2,6 +2,16 @@
+  - New feature flag `@aws-cdk/aws-iam:policyStatementValidateSid` enforces IAM SID validation
+  - SIDs must be alphanumeric (A-Z, a-z, 0-9) per AWS IAM requirements
+  - Validation occurs at synthesis time when flag is enabled
+  - Fixed invalid SIDs in aws-ecs cluster.ts to be alphanumeric
```
The CDK repository auto-generates `CHANGELOG.md` from conventional commit messages during the release process. Manual edits will cause merge conflicts and be overwritten on the next release.

Additionally, this line references "Fixed invalid SIDs in aws-ecs cluster.ts to be alphanumeric" but no such change exists in this PR's diff.

Per the [CONTRIBUTING guide](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md#step-5-commit), the changelog is generated automatically from the PR title - no manual entry is needed.

Please revert all changes to `CHANGELOG.md`.

alvazjor commented:
Thanks for working on this @camerondurham — catching invalid SIDs at synth time instead of deploy time is a real improvement. I have a suggestion that would simplify this PR significantly by removing the need for a feature flag entirely.

### The issue I see with the current approach

The validation currently lives in `PolicyDocument.resolve()`, which fires for **every** policy document — identity policies, resource policies, key policies, topic policies, etc. The problem is that the `[0-9A-Za-z]*` rule is specific to **IAM identity policies**. Other services have documented broader SID rules:

- **KMS** [explicitly documents](https://docs.aws.amazon.com/kms/latest/developerguide/key-policy-overview.html) that key policy SIDs **can include spaces**: *"The Sid element in a key policy statement can include spaces. (Spaces are prohibited in the Sid element of an IAM policy document.)"*
- **SNS** requires unique SIDs but documents no character restriction.
- **S3**, **SQS**, **Lambda** — no documented SID character restriction.

I suspect the feature flag was introduced because applying the validation universally would break resource policies that legitimately use broader SID characters (your prior PR #34828 had to modify existing integ tests for this reason, if I'm reading the history correctly). If that's the case, there's a simpler path that avoids the flag altogether.

### Proposed alternative: move validation to `validateForIdentityPolicy()`

CDK already has the infrastructure to distinguish identity policies from resource policies:

- `PolicyStatement.validateForIdentityPolicy()` is called from `Policy` (inline policies) and `ManagedPolicy` during synth via `node.addValidation()`
- `PolicyStatement.validateForResourcePolicy()` is called separately for resource policies

If you move the SID check into `validateForIdentityPolicy()`, it only fires where IAM actually enforces the rule. Non-alphanumeric SIDs in identity policies **always** fail at deploy time (IAM returns a 400), so this becomes a straight bug fix — no feature flag needed per our [contribution guidelines](https://github.com/aws/aws-cdk/blob/main/AGENTS.md#anti-patterns--things-not-to-do):

> Input that was accepted by synth but always failed at deploy time → feature flag NOT required; fail-fast synth-time validation is preferred.

The change would look like this in `policy-statement.ts`:

```ts
public validateForIdentityPolicy(): string[] {
  const errors = this.validateForAnyPolicy();
  // ... existing checks ...

  // IAM strictly enforces alphanumeric SIDs for identity policies
  if (this._sid !== undefined && !cdk.Token.isUnresolved(this._sid) && !/^[0-9A-Za-z]*$/.test(this._sid)) {
    errors.push(
      `Statement ID (sid) '${this._sid}' must be alphanumeric (A-Z, a-z, 0-9) when used in an identity-based policy. See https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_sid.html`,
    );
  }
  return errors;
}
```

### What this eliminates

- The feature flag definition in `cx-api/lib/features.ts`
- The exported constant (`IAM_POLICY_STATEMENT_VALIDATE_SID`)
- The `_validateSid()` method
- The flag check in `PolicyDocument.resolve()`
- All changes to `FEATURE_FLAGS.md` and `recommended-feature-flags.json`
- The flag context in the integ test

### Summary

The bug is real, the fix is welcome, but if it is possible to be scoped to where the rule actually applies, we wont need a feature flag at all. Happy to answer questions if anything is unclear.

camerondurham commented:
Great point @alvazjor , that's my bad I missed the SID restriction didn't apply to every type of policy document just the IAM ones.

I've moved SID validation to a dedicated function and removed the feature flag. I ran builds / new integ and framework tests but was not able to completely run deploy-mode on changes.

This PR should now just be scoped to **IAM Identity policies** only and the feature flag behavior removed.

Let me know if there's any additional feedback, thank you!

camerondurham commented on packages/aws-cdk-lib/aws-iam/test/policy-statement.test.ts:
```diff
@@ -260,4 +261,62 @@ describe('IAM policy statement', () => {
+      });
+
+      expect(statement.validateForIdentityPolicy()).toEqual([
+        "Statement ID (sid) 'Invalid-SID' must be alphanumeric (A-Z, a-z, 0-9) when used in an identity-based policy. See https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_sid.html",
```
This URL may change in the future and this would immediately become stale. Should this be removed? Are agents / llms good enough to look up and use latest AWS docs automatically so this is irreleant?

vishaalmehrishi replied:
I think it's fine either way. At this point there's more value in getting the change out so it benefits customers.
